GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,119 @@
+﻿using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace GVFS.Platform.Mac
+{
+    public class MacFileBasedLock : FileBasedLock
+    {
+        private const int InvalidFileDescriptor = -1;
+
+        private const int LockSh = 1; // #define LOCK_SH   1    /* shared lock */
+        private const int LockEx = 2; // #define LOCK_EX   2    /* exclusive lock */
+        private const int LockNb = 4; // #define LOCK_NB   4    /* don't block when locking */
+        private const int LockUn = 8; // #define LOCK_UN   8    /* unlock */
+
+        private int lockFileDescriptor;
+        private bool lockAcquired;
+
+        public MacFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath,
+            string signature)
+            : base(fileSystem, tracer, lockPath, signature) 
+        {
+            this.lockFileDescriptor = InvalidFileDescriptor;
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockAcquired)
+            {
+                return true;
+            }
+
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = Creat(this.LockPath, Convert.ToInt32("644", 8));
+                if (this.lockFileDescriptor == InvalidFileDescriptor)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+
+                    this.Tracer.RelatedWarning($"Failed to create lock file descriptor for '{this.LockPath}': {errno}"); 
+
+                    return false;
+                }
+            }
+
+            if (Flock(this.lockFileDescriptor, LockEx | LockNb) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                // Log error if not EWOULDBLOCK
+                this.Tracer.RelatedInfo($"Failed to acquire lock for '{this.LockPath}': {errno}");
+
+                return false;
+            }
+
+            byte[] signatureBytes = Encoding.UTF8.GetBytes(this.Signature);
+            long bytesWritten = Write(
+                this.lockFileDescriptor,
+                signatureBytes,
+                Convert.ToUInt64(signatureBytes.Length));
+            
+            if (bytesWritten == -1)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                this.Tracer.RelatedWarning($"Failed to write signature for '{this.LockPath}': {errno}");
+            }
+
+            this.Tracer.RelatedInfo($"Lock acquired for for '{this.LockPath}'");
+
+            return true;
+        }
+
+        public override void Dispose()
+        {
+            if (this.lockAcquired)
+            {
+                if (Flock(this.lockFileDescriptor, LockUn) != 0)
+                {
+                    this.Tracer.RelatedWarning($"Failed to release lock for: '{this.LockPath}'");

```

Comment:
Then maybe `IDisposable` isn't the proper interface for this class. Do we need a `TryUnlock` instead so we can retry?

Reply:
And should we avoid clearing `lockAcquired` when this happens?

Reply:
Reading through the comments for `flock` the lock will be released automatically when the file descriptor is closed, and so strictly speaking I don't think we need this call in the first place:

> Locks created by flock() are associated with an open file description
       (see open(2)).  This means that duplicate file descriptors (created
       by, for example, fork(2) or dup(2)) refer to the same lock, and this
       lock may be modified or released using any of these file descriptors.
       Furthermore, the lock is released either by an explicit LOCK_UN
       operation on any of these duplicate file descriptors, or when all
       such file descriptors have been closed.

http://man7.org/linux/man-pages/man2/flock.2.html

(Note also that all the error values described for `flock` only apply when acquiring the lock and not releasing it)

Reply:
Note also that if `close` fails it should not be retried, as the `fd` has been successfully released (any errors are just for diagnostic purposes):

> A careful programmer will check the return value of close(), since it
       is quite possible that errors on a previous write(2) operation are
       reported only on the final close() that releases the open file
       description.  Failing to check the return value when closing a file
       may lead to silent loss of data.  This can especially be observed
       with NFS and with disk quota.
>
>Note, however, that a failure return should be used only for
       diagnostic purposes (i.e., a warning to the application that there
       may still be I/O pending or there may have been failed I/O) or
       remedial purposes (e.g., writing the file once more or creating a
       backup).
>
>  Retrying the close() after a failure return is the wrong thing to do,
       since this may cause a reused file descriptor from another thread to
       be closed.  This can occur because the Linux kernel always releases
       the file descriptor early in the close operation, freeing it for
       reuse; the steps that may return an error, such as flushing data to
       the filesystem or device, occur only later in the close operation.

http://man7.org/linux/man-pages/man2/close.2.html